notes/force-move-implementation.md

```diff
@@ -0,0 +1,178 @@
+
+**Why not hash the `AppDefinition` together with the `ChannelId` into a `FixedPartHash`?** By doing this you perform one extra hash but you have to hash less data in the variable part. This tradeoff makes sense if you're hashing something like >3 variable parts for the same fixed part (based on the relative gas costs). We anticipate that most of the time channels will have 2 participants, so we optimize for this case.
+
+**Why not include `TurnNum` or `isFinal` in the `VariablePartHash`?** Having the `turnNum` and `isFinal` separate makes the `conclude` and `refute` methods more efficient, as they can pass in just the `VariablePartHash` and not the (potentially large) data within it.
+
+## Checking signatures
+
+Checking signatures is likely to be a signification part of the logic. Here's a potential algorithm:
+
+1. Let `nStates` be the number of states in the `states` array.
+   - Note that we assume that it has already been determined that these states form a chain of valid transitions and therefore must have incrementing turn numbers.
+2. Assume that `Signatures` takes the format of an array of length `nStates` where each element is an array of signatures on the corresponding state i.e. `sigs[i]` is an array of signatures on state `states[i]`.
+   - Note that it's possible for `states[i]` to be empty for some `i`.
+3. Let `nParticipants` be the number of participants.
+   - Note that we should have that `nStates <= nParticipants` and that the total number of signatures provided should be equal to `nParticipants`
+4. Let `hasSigned` be a boolean array of length `m` initialized to `false`.
+5. Let `maxTurnNum = states[states.length - 1].turnNum`.
+6. For `i = nStates - 1; i >= 0; i++`:
+   1. For `j = 0; j < signatures[i].length; j++`:
+      1. Determine that `signatures[i][j]` is a valid signature for some participant on state `states[i]`. Let `p` be the index of this participant in the channel's particpants array.
+      1. Set the bit `hasSigned[(p - maxTurnNum - 1) % n]`.
+   2. Fail unless `hasSigned[i - nStates + nParticipants]` is set.
+7. Fail unless `hasSigned[0..(nParticipants - nStates - 1)]` are all set.
+8. Return true
+
+**Can we optimize this further by changing the way states / sigs are passed in?** For example passing the arrays in last-state-first might help.
+
+## Updates call signatures
+
+[TODO] Rewrite the method call signatures to avoid any unnecessary repetition in the data being passed in.
+
+## Unanswered Questions
+
+1. Should we use time or blocknumber? Time is generally more predictable and nicer to work with in client code but is manipulable by miners :(.
+2. Is it worth optimizing separately for the case n=2 (e.g. with methods `forceMove2` and so on)
+3. What order should we put states in? (Maybe reverse?)
+4. Should we use abi encode?
+5. We don't technically need to pass the `isFinal` flag into `conclude` - it will only work if it is true. Maybe that's a step too far?
```

Comment:
In my opinion this is a code complexity vs efficiency tradeoff. I think code simplicity is valuable.